sys/riscv/riscv/pmap.c

```diff
@@ -3603,6 +3639,464 @@ pmap_enter_quick_locked(pmap_t pmap, vm_offset_t va, vm_page_t m,
+		 * to find the page in question.
+		 *
+		 * XXX We'd rather just demote the L2 page right now, surely?
+		 */
+		m = NULL;
+		res = PMAP_CAPLOADGEN_UNABLE;
+		goto out;
+	}
+	pte = pmap_l2_to_l3(l2, va);
+	if (pte == NULL || ((oldpte = pmap_load(pte)) & PTE_V) == 0) {
+		m = NULL;
+		res = PMAP_CAPLOADGEN_UNABLE;
+		goto out;
+	}
+
+	switch (oldpte & (PTE_CR | PTE_CRM)) {
+	case 0:		/* tag clearing */
+	case PTE_CRM:	/* always trapping; not something we can fix? */
+
+		/*
+		 * TODO This seems wrong.  I guess the good news is that we
+		 * probably don't hit these cases often, but when we do it's
+		 * probably for IDLE pages and we should just return m and let
+		 * the system see that PGA_CAPSTORE is clear.  Alternatively,
+		 * the page has become PGA_CAPSTORE while we haven't been
+		 * looking?
+		 */
+
+		m = NULL;
+		res = PMAP_CAPLOADGEN_UNABLE;
+		goto out;
+	case PTE_CR:	/* always allowed; not for us to worry about? */
+		panic("Unexpected PTE @ CR !CRM; revocation not optional yet");
+	case PTE_CR | PTE_CRM: /* ah, here we go */
+		break;
+	}
+
+	if (!(oldpte & PTE_CRG) == !(pmap->flags.uclg)) {
+		/* Page already scanned, just fence (maybe redundantly) */
+		if (flags & PMAP_CAPLOADGEN_UPDATETLB)
+			sfence_vma_page(va);
+
+		m = NULL;
+		res = PMAP_CAPLOADGEN_ALREADY;
+		goto out;
+	}
+
+	KASSERT(oldpte & PTE_U, ("!PTE_U w/ CLG mismatch va=%lx", va));
+
+	m = PHYS_TO_VM_PAGE(PTE_TO_PHYS(oldpte));
+	if (*mp == m) {
+		/*
+		 * We expected this page here (i.e., this is the page we just
+		 * scanned), so go ahead and update.  We know that the CLG bits
+		 * must still be wrong in light of the earlier test.
+		 */
+		res = PMAP_CAPLOADGEN_OK;
+
+		if (!(flags & PMAP_CAPLOADGEN_HASCAPS)) {
+			/*
+			 * We didn't see a capability on this page; step this
+			 * PTE closer to being cap-clean.
+			 */
+			if (oldpte & PTE_CD) {
+				/*
+				 * CAP-DIRTY -> DIRTYABLE; we just do a store
+				 * here as there might be a capability store
+				 * between us having examined the page and the
+				 * store-release below, and so we're going to
+				 * scan the page again anyway.
+				 */
+				pmap_clear_bits(pte, PTE_CD);
+			} else if (oldpte & PTE_CW) {
+				/* PTE CAP-DIRTYABLE -> CAP-CLEAN? */
+				pmap_caploadgen_update_clear_cw(pte, oldpte);
+			} else if (flags & PMAP_CAPLOADGEN_NONEWMAPS) {
+				/* No new mappings possible */
+				vm_page_astate_t mas = vm_page_astate_load(m);
+
+				if (mas.flags & PGA_CAPDIRTY) {
+					/* Page-level cleaning: -?> VACANT */
+					vm_page_aflag_clear(m, PGA_CAPDIRTY);
+				} else {
+					KASSERT(mas.flags & PGA_CAPSTORE,
+					    ("Page already CAP-CLEAN?"));
+					/* PTE CAP-CLEAN; page -?> IDLE */
+
+					/* XXXNWF Why? */
+					if (flags & PMAP_CAPLOADGEN_UPDATETLB) {
+						sfence_vma_page(va);
+					}
+					PMAP_UNLOCK(pmap);
+					/*
+					 * XXXMJ don't we need to hold a
+					 * reference to m?  What prevents
+					 * concurrent reclamation?
+					 */
```

Comment:
I believe `NONEWMAPS` implies that we have `m` `xbusy`.  See https://github.com/CTSRD-CHERI/cheribsd/pull/1857/files#diff-f2fa36d86e8ebacf87f36a65bec6148f7a5f7fefdbf630531d923c09907cf2a2R612